torch_xla/csrc/runtime/ifrt_computation_client.cc

```diff
@@ -0,0 +1,643 @@
+  // tracked in XlaCoordinator, so the PjRtClient must be destroyed first.
+  client_ = nullptr;
+  coordinator_ = nullptr;
+}
+
+bool IfrtComputationClient::CoordinatorInitialized() const {
+  return coordinator_ != nullptr;
+}
+
+void IfrtComputationClient::InitializeCoordinator(int global_rank,
+                                                  int world_size,
+                                                  std::string master_addr,
+                                                  std::string port) {
+  XLA_CHECK(coordinator_ == nullptr)
+      << "Can only initialize the XlaCoordinator once.";
+  coordinator_ = std::make_unique<XlaCoordinator>(global_rank, world_size,
+                                                  master_addr, port);
+}
+
+XlaCoordinator& IfrtComputationClient::GetCoordinator() {
+  XLA_CHECK(coordinator_ != nullptr)
+      << "XlaCoordinator has not been initialized";
+  return *coordinator_;
+}
+
+void IfrtComputationClient::IfrtData::Assign(
+    const torch::lazy::BackendData& data) {
+  const IfrtData& pjrt_data = dynamic_cast<const IfrtData&>(data);
+  if (&pjrt_data != this) {
+    buffer = pjrt_data.buffer;
+  }
+}
+
+xla::OpSharding IfrtComputationClient::IfrtData::GetSharding() const {
+  XLA_CHECK(HasSharding()) << "Check HasSharding first";
+  return *sharding_;
+}
+
+ComputationClient::DataPtr IfrtComputationClient::CreateDataPlaceholder(
+    std::string device, xla::Shape shape) {
+  return std::make_shared<IfrtData>(device, shape);
+}
+
+std::vector<ComputationClient::DataPtr> IfrtComputationClient::GetDataShards(
+    ComputationClient::DataPtr data) {
+  tsl::profiler::TraceMe activity("IfrtComputationClient::GetDataShards",
+                                  tsl::profiler::TraceMeLevel::kInfo);
+  std::vector<ComputationClient::DataPtr> shards;
+  if (data->HasSharding()) {
+    auto ifrt_data = std::dynamic_pointer_cast<IfrtData>(data);
+    std::vector<tsl::RCReference<xla::ifrt::Array>> arrays =
+        ifrt_data->buffer
+            ->DisassembleIntoSingleDeviceArrays(
+                xla::ifrt::ArrayCopySemantics::kAlwaysCopy)
+            .value();
+
+    for (auto array : arrays) {
+      shards.push_back(std::make_shared<IfrtData>(
+          PjRtDeviceToString(array->sharding().devices()[0]), array));
+    }
+  } else {
+    shards.push_back(data);
+  }
+  return shards;
+}
+
+ComputationClient::DataPtr IfrtComputationClient::GetDataShard(
+    ComputationClient::DataPtr data, size_t index) {
+  tsl::profiler::TraceMe activity("IfrtComputationClient::GetDataShard",
+                                  tsl::profiler::TraceMeLevel::kInfo);
+  return GetDataShards(data)[index];
+}
+
+ComputationClient::DataPtr IfrtComputationClient::WrapDataShards(
+    const std::vector<DataPtr>& shards, std::string device, xla::Shape shape,
+    xla::OpSharding sharding) {
+  // TODO: implement CreateDataPlaceholder for sharded data
+  if (shards.size() == 0) {
```

Comment:
is there a legit use case of `shards.zie() == 0`?

Reply:
This is how sharded data placeholders get created right now:

https://github.com/pytorch/xla/blob/a2f80e466aedd7f1457bda75e9beea9c76509f1b/torch_xla/csrc/xla_sharding_util.cpp#L590-L593

I'd rather update `CreateDataPlaceholder` to take a sharding and make it more explicit